From e2c8119e5e999f4978a60a0ce80f76b13b350456 Mon Sep 17 00:00:00 2001
From: "Jasper St. Pierre" <jstpierre@mecheye.net>
Date: Mon, 6 Jun 2011 07:54:19 -0400
Subject: [PATCH] Update ropemacs backend.

---
 auto-complete-config.el |   87 ++++++++++++++++++++++++++++------------------
 1 files changed, 53 insertions(+), 34 deletions(-)

diff --git a/auto-complete-config.el b/auto-complete-config.el
index a12b6f2..1ee9cec 100644
--- a/auto-complete-config.el
+++ b/auto-complete-config.el
@@ -409,44 +409,61 @@
 
 
 
-;;;; Not maintained sources
-
 ;; ropemacs
 
-(defvar ac-ropemacs-loaded nil)
-(defun ac-ropemacs-require ()
-  (with-no-warnings
-    (unless ac-ropemacs-loaded
-      (pymacs-load "ropemacs" "rope-")
-      (if (boundp 'ropemacs-enable-autoimport)
-          (setq ropemacs-enable-autoimport t))
-      (setq ac-ropemacs-loaded t))))
+(defun ac-ropemacs-candidates ()
+  (mapcar (lambda (completion)
+      (concat ac-prefix completion))
+    (rope-completions)))
+
+(ac-define-source nropemacs
+  '((candidates . ac-ropemacs-candidates)
+    (symbol     . "p")))
+
+(ac-define-source nropemacs-dot
+  '((candidates . ac-ropemacs-candidates)
+    (symbol     . "p")
+    (prefix     . c-dot)
+    (requires   . 0)))
+
+(defun ac-nropemacs-setup ()
+  (setq ac-sources (append '(ac-source-nropemacs
+                             ac-source-nropemacs-dot) ac-sources)))
+
+;; extended ropemacs
+
+(defun ac-eropemacs-candidates ()
+  (mapcar (lambda (proposal)
+          (destructuring-bind (name doc type) proposal
+            (list (concat ac-prefix name) doc
+                  (if type (substring type 0 1) nil))))
+        (rope-extended-completions)))
+
+(defun ac-eropemacs-document (item) (car  item))
+(defun ac-eropemacs-symbol   (item) (cadr item))
+
+(ac-define-source extended-ropemacs
+  '((candidates . ac-eropemacs-candidates)
+    (document   . ac-eropemacs-document)
+    (symbol     . ac-eropemacs-symbol)))
+
+(ac-define-source extended-ropemacs-dot
+  '((candidates . ac-eropemacs-candidates)
+    (document   . ac-eropemacs-document)
+    (symbol     . ac-eropemacs-symbol)
+    (prefix     . c-dot)
+    (requires   . 0)))
+
+(defun ac-eropemacs-setup ()
+  (setq ac-sources (append '(ac-source-extended-ropemacs
+                             ac-source-extended-ropemacs-dot) ac-sources)))
 
 (defun ac-ropemacs-setup ()
-  (ac-ropemacs-require)
-  ;(setq ac-sources (append (list 'ac-source-ropemacs) ac-sources))
-  (setq ac-omni-completion-sources '(("\\." ac-source-ropemacs))))
-
-(defun ac-ropemacs-initialize ()
-  (autoload 'pymacs-apply "pymacs")
-  (autoload 'pymacs-call "pymacs")
-  (autoload 'pymacs-eval "pymacs" nil t)
-  (autoload 'pymacs-exec "pymacs" nil t)
-  (autoload 'pymacs-load "pymacs" nil t)
-  (add-hook 'python-mode-hook 'ac-ropemacs-setup)
-  t)
-
-(defvar ac-ropemacs-completions-cache nil)
-(defvar ac-source-ropemacs
-  '((init
-     . (lambda ()
-         (setq ac-ropemacs-completions-cache
-               (mapcar
-                (lambda (completion)
-                  (concat ac-prefix completion))
-                (ignore-errors
-                  (rope-completions))))))
-    (candidates . ac-ropemacs-completions-cache)))
+  (if (functionp 'rope-extended-completions)
+      (add-hook 'python-mode-hook 'ac-eropemacs-setup)
+    (add-hook 'python-mode-hook 'ac-nropemacs-setup)))
+
+;;;; Not maintained sources
 
 ;; rcodetools
 
@@ -490,7 +507,9 @@
   (add-hook 'ruby-mode-hook 'ac-ruby-mode-setup)
   (add-hook 'css-mode-hook 'ac-css-mode-setup)
   (add-hook 'auto-complete-mode-hook 'ac-common-setup)
+  (add-hook 'python-mode 'ac-ropemacs-setup)
   (global-auto-complete-mode t))
 
 (provide 'auto-complete-config)
 ;;; auto-complete-config.el ends here
+
-- 
1.7.6

